lib/GenomeFileUtil/core/GenomeUtils.py

```diff
@@ -369,3 +377,107 @@ def sort_dict(in_struct):
+    This will not override any preset or user-passed fields.
+    Mutates genome_dict.
+    """
+    sciname = genome_dict.get('sciname')
+    if sciname:
+        genome_dict.setdefault('taxonomy', f'Unconfirmed Organism: {sciname}')
+    else:
+        genome_dict.setdefault('taxonomy', 'Unconfirmed Organism')
+    # XXX genetic_code should probably not have a default, but this matches previous/legacy behavior.
+    genome_dict.setdefault('genetic_code', 11)
+    genome_dict.setdefault('domain', 'Unknown')
+
+
+def set_taxon_data(tax_id, re_api_url, genome_dict):
+    """
+    Fetch and set taxonomy data for a genome using an NCBI taxonomy ID.
+
+    We mutate the genome_dict with the following fields:
+    {
+      "taxonomy": "x;y;z",    # NCBI taxonomy lineage string for human readability
+      "domain": "x"           # String name of the domain
+      "genetic_code": 11      # NCBI categorization of the lineage
+                               (https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi)
+      "taxon_assignments": {  # Mapping of taxonomy namespace to taxonomy ID
+        "ncbi": 1234
+      }
+    }
+    """
+    # We use the Relation Engine to do a lookup on taxonomy data from ID
+    re_client = REClient(re_api_url)
+    # FIXME this timestamp should come from the client
+    now = int(time.time() * 1000)  # unix epoch for right now, for use in the RE API
+    tax_id = str(tax_id)
+    genome_dict.setdefault('warnings', [])
+    assignments = genome_dict.get('taxon_assignments')
+    # Check to make sure that tax_id == genome.taxon_assignments.ncbi
+    if assignments and assignments.get(_NCBI_TAX) and assignments[_NCBI_TAX] != tax_id:
+        raise RuntimeError(
+            f"The provided taxon ID ({tax_id}) differs from the "
+            f"taxon ID in the genome's `taxon_assignments` field: {assignments[_NCBI_TAX]}."
+        )
+    genome_dict['taxon_assignments'] = {'ncbi': tax_id}
+    # Fetch the taxon from Relation Engine by taxon ID
+    try:
+        resp_json = re_client.stored_query(
+            'ncbi_fetch_taxon',
+            {'id': str(tax_id), 'ts': now},
+            raise_not_found=True)
+    except RENotFound as err:
+        # Taxon not found; log and raise
+        logging.error(str(err))
+        raise err
+    re_result = resp_json['results'][0]
+    # Refer to the following schema for returned fields in `re_result`:
+    # https://github.com/kbase/relation_engine_spec/blob/develop/schemas/ncbi/ncbi_taxon.yaml
+    gencode = int(re_result['gencode'])
+    # If there is a mismatch on some of these fields from NCBI, save a warning and continue
+    if genome_dict.get('genetic_code') and genome_dict['genetic_code'] != gencode:
+        genome_dict['warnings'].append(
+            f"The genetic code provided by NCBI ({gencode}) "
+            f"does not match the one given by the user ({genome_dict['genetic_code']})"
+        )
+    genome_dict['genetic_code'] = gencode
```

Comment:
do we want to log if there is no genetic code in the genome_dict initially as a warning?

Reply:
Maybe?